clib/stream.ml

```diff
@@ -0,0 +1,184 @@
+
+(* We use exception Foo = Stdlib.Stream.Foo to make it easier for
+   plugins since they won't be getting type errors when using them. *)
+
+exception Failure = Stdlib.Stream.Failure
+[@@ocaml.warning "-3"]
+
+exception Error = Stdlib.Stream.Error
+[@@ocaml.warning "-3"]
+
+let count { count } = count
+
+let fill_buff b =
+  b.len <- input b.ic b.buff 0 (Bytes.length b.buff); b.ind <- 0
+
+
+let rec get_data : type v. v data -> v data = fun d -> match d with
+ (* Returns either Sempty or Scons(a, _) even when d is a generator
+    or a buffer. In those cases, the item a is seen as extracted from
+ the generator/buffer.
+ The count parameter is used for calling `Sgen-functions'.  *)
+   Sempty | Scons (_, _) -> d
+ | Sapp (d1, d2) ->
+     begin match get_data d1 with
+       Scons (a, d11) -> Scons (a, Sapp (d11, d2))
+     | Sempty -> get_data d2
+     | _ -> assert false
+     end
+ | Sgen {curr = Some None} -> Sempty
+ | Sgen ({curr = Some(Some a)} as g) ->
+     g.curr <- None; Scons(a, d)
+ | Sgen g ->
+     begin match g.func () with
+       None -> g.curr <- Some(None); Sempty
+     | Some a -> Scons(a, d)
+         (* Warning: anyone using g thinks that an item has been read *)
+     end
+ | Sbuffio b ->
+     if b.ind >= b.len then fill_buff b;
+     if b.len == 0 then Sempty else
+       let r = Bytes.unsafe_get b.buff b.ind in
+       (* Warning: anyone using g thinks that an item has been read *)
+       b.ind <- succ b.ind; Scons(r, d)
+
+
+let peek : type v. v t -> v option = fun s ->
+ (* consult the first item of s *)
+ match s.data with
+   Sempty -> None
+ | Scons (a, _) -> Some a
+ | Sapp (_, _) ->
+     begin match get_data s.data with
+       Scons(a, _) as d -> s.data <- d; Some a
+     | Sempty -> None
+     | _ -> assert false
+     end
+ | Sgen {curr = Some a} -> a
+ | Sgen g -> let x = g.func () in g.curr <- Some x; x
+ | Sbuffio b ->
+     if b.ind >= b.len then fill_buff b;
+     if b.len == 0 then begin s.data <- Sempty; None end
+     else Some (Bytes.unsafe_get b.buff b.ind)
+
+
+let rec junk : type v. v t -> unit = fun s ->
+  match s.data with
+    Scons (_, d) -> s.count <- (succ s.count); s.data <- d
+  | Sgen ({curr = Some _} as g) -> s.count <- (succ s.count); g.curr <- None
+  | Sbuffio b ->
+      if b.ind >= b.len then fill_buff b;
+      if b.len == 0 then s.data <- Sempty
+      else (s.count <- (succ s.count); b.ind <- succ b.ind)
+  | _ ->
+      match peek s with
+        None -> ()
+      | Some _ -> junk s
+
+let rec nget n s =
+  if n <= 0 then [], s.data, 0
+  else
+    match peek s with
+      Some a ->
+        junk s;
+        let (al, d, k) = nget (pred n) s in a :: al, Scons (a, d), succ k
+    | None -> [], s.data, 0
+
+
+let npeek n s =
+  let (al, d, len) = nget n s in
+  s.count <- (s.count - len);
+  s.data <- d;
+  al
+
+let nth n st =
+  try List.nth (npeek (n+1) st) n
+  with Stdlib.Failure _ -> raise Failure
+
+let rec njunk n st =
```

Comment:
Both njunk and npeek are unused.

Reply:
npeek is used in a bunch of places (18 matches for Stream.npeek in grep)